clang/runtime/dpct-rt/include/dpl_extras/algorithm.h

```diff
@@ -1746,6 +1746,87 @@ segmented_reduce_argmax(Policy &&policy, Iter1 keys_in, Iter2 keys_out,
+    OutputIterator3 num_runs) {
+  using oneapi::dpl::make_transform_iterator;
+  using oneapi::dpl::make_zip_iterator;
+  using offsets_t =
+      typename ::std::iterator_traits<OutputIterator1>::value_type;
+  using lengths_t =
+      typename ::std::iterator_traits<OutputIterator2>::value_type;
+
+  auto n = ::std::distance(input_beg, input_end);
+  // First element must be nontrivial run (start of first segment)
+  auto first_adj_it = oneapi::dpl::adjacent_find(policy, input_beg, input_end);
+  auto first_adj_idx = ::std::distance(input_beg, first_adj_it);
+  if (first_adj_it == input_end) {
+    ::std::fill(policy, num_runs, num_runs + 1, 0);
+    return ::std::make_pair(offsets_out, lengths_out);
+  }
+  auto get_prev_idx_element = [first_adj_idx](const auto &idx) {
+    auto out_idx = idx + first_adj_idx;
+    return (out_idx == 0) ? 0 : out_idx - 1;
+  };
+  auto get_next_idx_element = [first_adj_idx, n](const auto &idx) {
+    auto out_idx = idx + first_adj_idx;
+    return (out_idx == n - 1) ? n - 1 : out_idx + 1;
+  };
+  // TODO: Use shifted view to pad range once oneDPL ranges is non-experimental
+  auto left_shifted_input_beg =
+      oneapi::dpl::make_permutation_iterator(input_beg, get_prev_idx_element);
+  auto right_shifted_input_beg =
+      oneapi::dpl::make_permutation_iterator(input_beg, get_next_idx_element);
+  // Segment type for ith idx consists of zip of iterators at (i-1, i, i+1)
+  // padded at the ends
+  auto zipped_keys_beg = make_zip_iterator(
+      left_shifted_input_beg, input_beg, right_shifted_input_beg,
+      oneapi::dpl::counting_iterator<offsets_t>(0));
+  // Set flag at the beginning of new nontrivial run (ex: (0, 1, 1) -> 1)
+  auto key_flags_beg =
+      make_transform_iterator(zipped_keys_beg, [n](const auto &zipped) {
+        using ::std::get;
+        bool zero_idx_mask = get<3>(zipped) != 0;
+        bool last_idx_mask = get<3>(zipped) != n - 1;
+        return (zero_idx_mask * (get<0>(zipped) != get<1>(zipped)) &&
+                get<1>(zipped) == get<2>(zipped)) *
+               last_idx_mask;
+      });
+  auto count_beg = oneapi::dpl::counting_iterator<offsets_t>(0);
+  auto const_it = dpct::make_constant_iterator(lengths_t(1));
+  // Check for presence of nontrivial element at current index
+  auto tr_nontrivial_flags = make_transform_iterator(
+      make_zip_iterator(left_shifted_input_beg, input_beg,
+                        right_shifted_input_beg),
+      [](const auto &zip) {
+        using ::std::get;
+        return get<0>(zip) == get<1>(zip);
+      });
+  auto zipped_vals_beg =
+      make_zip_iterator(tr_nontrivial_flags, count_beg, const_it);
+  auto pred = [](bool lhs, bool rhs) { return !rhs; };
+  // Update count of run. Starting index of run stored in assignment of initial
+  // element (first element in each segment).
+  auto op = [](auto lhs, const auto &rhs) {
+    using ::std::get;
+    get<2>(lhs) += get<0>(rhs);
```

Comment:
I think a comment here about how the `lhs` count index (`get<1>`) is preserved and passed through to the output here would be helpful.

Reply:
I have added a comment to explain all get<1> holds the start of each run and is preserved throughout each segment's reduction.